sample/trace_key.go

```diff
@@ -40,75 +43,194 @@ func newTraceKey(fields []string, useTraceLength bool) *traceKey {
+		d.keyBuilder.WriteString(strconv.FormatInt(int64(len(spans)), 10))
 		fieldCount += 1
 	}
 
-	return key.String(), fieldCount
+	return d.keyBuilder.String(), fieldCount
+}
+
+// distinctValue keeps track of distinct values for a set of fields.
+// It stores the unique values as strings.
+type distinctValue struct {
+	buf          []byte
+	fields       []string
+	values       []map[uint64]string
+	valuesBuffer []string
+
+	// totalUniqueCount keeps track of how many unique values we've seen so far for a trace key.
+	totalUniqueCount int
+	// maxDistinctValue is the maximum number of distinct values we will store for a trace key.
+	maxDistinctValue int
+}
+
+func (d *distinctValue) init(fields []string, maxDistinctValue int) {
+	for i := range fields {
+		if i >= len(d.fields) {
+			// if we don't have enough fields, allocate more
+			d.fields = append(d.fields, fields[i])
+			continue
+		}
+		d.fields[i] = fields[i]
+	}
+
+	if len(d.values) < len(fields) {
+		// if we don't have enough values, allocate more
+		for i := len(d.values); i < len(fields); i++ {
+			d.values = append(d.values, make(map[uint64]string))
+		}
+	} else {
+		// if we have more values than fields, trim the excess
+		d.values = d.values[:len(fields)]
+	}
+
+	d.maxDistinctValue = maxDistinctValue
+}
+
+func (d *distinctValue) Reset() {
+	// Reset the fields and values but do not reallocate them
+	for i := range d.fields {
+		d.fields[i] = ""
+	}
+	for i := range d.values {
+		if d.values[i] != nil {
+			for k := range d.values[i] {
+				delete(d.values[i], k)
+			}
+		}
+	}
+
+	// Reset the total unique count
+	d.totalUniqueCount = 0
+	// Reset the buffer
+	d.buf = d.buf[:0]
+	// Reset the values buffer
+	d.valuesBuffer = d.valuesBuffer[:0]
+}
+
+// Values returns the distinct values for a given field index.
+// It returns a sorted slice of strings containing the unique values for that field.
+func (d *distinctValue) Values(fieldIdx int) []string {
+	if fieldIdx < 0 || fieldIdx >= len(d.values) {
+		return nil
+	}
+
+	// Get the map for the specified field index
+	valueMap := d.values[fieldIdx]
+	if len(valueMap) == 0 {
+		return nil
+	}
+
+	// use the valuesBuffer to avoid unnecessary allocations
+	d.valuesBuffer = d.valuesBuffer[:0]
+	for _, value := range valueMap {
+		d.valuesBuffer = append(d.valuesBuffer, value)
+	}
+	sort.Strings(d.valuesBuffer)
+
+	return d.valuesBuffer
+}
+
+// AddAsString adds a value to the distinct values for a given field index.
+// It returns true if the value was added, false if it was already present or if the maxDistinctValue limit was reached.
+func (d *distinctValue) AddAsString(value any, fieldIdx int) bool {
+	if value == nil {
+		return false
+	}
+
+	d.buf = d.buf[:0] // reset the buffer for each new value
+
+	switch v := value.(type) {
+	case string:
+		d.buf = append(d.buf, []byte(v)...)
+	case int:
+		d.buf = strconv.AppendInt(d.buf, int64(v), 10)
+	case int64:
+		d.buf = strconv.AppendInt(d.buf, v, 10)
```

Comment:
I _think_ you can do this in one case., casting an int64 to an int64 is allowed.
```
case int, int64:
```